src/app/clusters/bindings/BindingManager.h

```diff
@@ -0,0 +1,173 @@
+ *
+ * The connection is managed by the stack and peer_device is guaranteed to be available.
+ * The application shall decide the content to be sent to the peer.
+ *
+ * E.g. The application will send on/off commands to peer for the OnOff cluster.
+ *
+ */
+using BoundDeviceChangedHandler = void (*)(EndpointId localEndpoint, EndpointId remoteEndpoint, ClusterId clusterId,
+                                           DeviceProxy * peer_device, Optional<GroupId> group);
+
+/**
+ *
+ * The BindingManager class manages the CASEConnection and attribute writes to bound peers.
+ *
+ * The binding cluster will tell the device about the peer to connect to and the cluster to watch.
+ * A CASE connection will be triggered when:
+ *  - The binding cluster adds an entry to the binding table.
+ *  - A watched cluster changes but we cannot find an active connection to the peer.
+ *
+ * The class uses an LRU mechanism to choose the connection to eliminate when there is no space for a new connection.
+ * The BindingManager class will not actively re-establish connection and will connect on-demand (when binding cluster
+ * or watched cluster is changed).
+ *
+ */
+class BindingManager
+{
+public:
+    BindingManager() :
+        mOnConnectedCallback(HandleDeviceConnected, this), mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+    {}
+
+    void RegisterBoundDeviceChangedHandler(BoundDeviceChangedHandler handler) { mBoundDeviceChangedHandler = handler; }
+
+    void SetAppServer(Server * appServer) { mAppServer = appServer; }
+
+    /*
+     * Adds a binding entry. Can be called for adding a new cluster to a node already connected.
+     *
+     */
+    CHIP_ERROR CreateBinding(FabricIndex fabric, NodeId node, EndpointId localEndpoint, ClusterId cluster);
+
+    /*
+     * Totally disconnect a device. Call this function only when no binding table entry is associated with the node.
+     *
+     */
+    CHIP_ERROR DisconnectDevice(FabricIndex fabric, NodeId node);
+
+    /*
+     * Notify a cluster change to **all** bound devices associated with the cluster.
+     *
+     */
+    CHIP_ERROR NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster);
+
+    static BindingManager & GetInstance() { return sBindingManager; }
+
+private:
+    static BindingManager sBindingManager;
+
+    static constexpr uint8_t kMaxPendingClusters = 5;
+
+    struct ClusterPath
+    {
+        EndpointId endpoint;
+        ClusterId cluster;
+    };
+
+    // A pending command to be sent to a bound peer waiting for the CASE session to be established.
+    class PendingClusterEntry
+    {
+    public:
+        PendingClusterEntry(PeerId peerId) : mPeerId(peerId) {}
+
+        PeerId GetPeerId() { return mPeerId; }
+
+        System::Clock::Timestamp GetLastUpdateTime() { return mLastUpdateTime; }
+        void Touch() { mLastUpdateTime = System::SystemClock().GetMonotonicTimestamp(); }
```

Comment:
Currently the CASESessionManager does not actively close the sessions and we need to move the LRU mechanism to the CASESessionManager to avoid breaking sessions used somewhere else. TODO added.